### PR TITLE
Inline manifest.json

### PR DESCRIPTION
**Pull request by silverwind**
Improve performance by eliminating this separate request and just inline this small JSON in HTML directly as a data uri. Also, I've updated the previously static app name scripts to use `AppName` from config. `theme_color` is removed because it's duplicating with the existing meta tag.

I've confirmed this as working via "Add to Homescreen" feature which offered to save the shortcut under the new app name.

### CODE REVIEW

**Comment by zeripath** [comment]
Is the data uri always going to work?

**Comment by silverwind** [comment]
Yes, browsers are pretty much forced to implement data URIs if they want to conform.

We just need to be careful with data emitted in the string because it must be URL-escaped. I guess `StaticUrlPrefix` and `AppSubUrl` should generally only contain URL-safe characters but I guess we could also just escape them.

**Comment by silverwind** [comment]
Okay everything is escaped now including the URLs. Browsers will URL-decode the string before parsing it is it's safe to do in all cases.

File: main.go [hunk 1]
```diff
@@ -41,7 +40,6 @@ var (
 func init() {
 	setting.AppVer = Version
 	setting.AppBuiltWith = formatBuiltWith()
-	setting.AppStartTime = time.Now().UTC()
```
Comment (techknowlogick) [inline main.go]: This change seems unrelated, could you describe why this is being removed?

Comment (silverwind) [reply main.go]: I's becoming unused because the `/manifest.json` route was the only consumer of the variable (I added it with the HTTP cache commit). We can keep it around if you prefer.

**Comment by zeripath** [comment]
Is it possible keep the file but insert its contents directly into the attribute?

That way you can keep the endpoint around for those who don't update their templates

**Comment by silverwind** [comment]
Can't insert the JSON as-is because it needs URL escaping and I'm generally leaning to not do for the sake of performance.

What is the issue exactly about people not updating their templates? I thought templates work by the way of inserting user content at strategic places in the HTML. How come a user would have these parts hardcoded in their template customization?

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
It could be put in `bindata` and load on template rendering if necessary.

**Comment by silverwind** [comment]
I was thinking about generating it once on app startup and caching it in memory for even more speed, thought I guess the gains would be minimal.

File: main.go (hunk 1 above)
Comment (a1012112796) [reply main.go]: Hmm, I think it's duplicate with this line also.
https://github.com/go-gitea/gitea/blob/4aabbacdfe2830e3b3808e42d6d89d6688754089/routers/admin/admin.go#L43

**Review by a1012112796 — APPROVED** [review]

**Comment by lunny** [comment]
Please resolve the conflicts.

File: main.go (hunk 1 above)
Comment (silverwind) [reply main.go]: I've restored `setting.AppStartTime` and it's now being used on the admin page:

image

File: modules/templates/helper.go [hunk 2]
```diff
@@ -88,6 +88,9 @@ func NewFuncMap() []template.FuncMap {
 		"AllowedReactions": func() []string {
 			return setting.UI.Reactions
 		},
+		"ManifestData": func() template.URL {
```
Comment (lunny) [inline modules/templates/helper.go]: Maybe we should set this as a global var but not use a function. I think Golang will not optimize here.

Comment (silverwind) [reply modules/templates/helper.go]: Hmm sure if I find out how. I guess the same goes for the other `setting` variables above as well.

Comment (silverwind) [reply modules/templates/helper.go]: I'm not sure where we'd add such variables. [`ExecuteTemplate`](https://golang.org/pkg/html/template/#Template.ExecuteTemplate) is only called in macaron renderer and that one does not seem to allow additional static variables on instantiation (https://godoc.org/gopkg.in/macaron.v1#RenderOptions).

Comment (lunny) [reply modules/templates/helper.go]: It's not obvious. It's in https://github.com/go-gitea/gitea/blob/master/modules/context/context.go#L346. And you are right. All inline functions should be moved there.

Comment (silverwind) [reply modules/templates/helper.go]: Indeed, but let's do that in another PR.

Comment (silverwind) [reply modules/templates/helper.go]: Did it for this variable now.

Comment (silverwind) [reply modules/templates/helper.go]: Thought I have to say there is one benefit of using a helper vs. a variable. Helper can be used anywhere easily, even withing a `range` subtemplate which would otherwise require awkward passing of data using the `dict` helper. So I think it's fine to do in this case, but I prefer using the helper style in such subtemplate cases.

**Comment by wULLSnpAXbWZGYDYyhWTKKspEQoaYxXyhoisqHf** [comment]
> What is the issue exactly about people not updating their templates? I thought templates work by the way of inserting user content at strategic places in the HTML. How come a user would have these parts hardcoded in their template customization?

I am keeping around a custom template.
What exactly would I have to update in my `pwa/manifest_json.tmpl` to make it work once this is merged?
From what I gathered I should be ok :D

**Comment by silverwind** [comment]
Ah, so you are customizing the "private" templates, didn't know that was possible.

You can just delete it, it doesn't exist anymore after this change.

**Comment by wULLSnpAXbWZGYDYyhWTKKspEQoaYxXyhoisqHf** [comment]
> Ah, so you are customizing the "private" templates, didn't know that was possible.
> 
> You can just delete it, it doesn't exist anymore after this change.

well that's too bad because I'm doing a little customization in there, such as using relative `/img` references.

File: modules/templates/helper.go (hunk 2 above)
Comment (lunny) [reply modules/templates/helper.go]: Hm, you are right. No global variables but global functions on golang template.

**Comment by lunny** [comment]
> > Ah, so you are customizing the "private" templates, didn't know that was possible.
> > You can just delete it, it doesn't exist anymore after this change.
> 
> well that's too bad because I'm doing a little customization in there, such as using relative `/img` references.

I think you could find other way to do that on Gitea.

**Comment by wULLSnpAXbWZGYDYyhWTKKspEQoaYxXyhoisqHf** [comment]
> > > Ah, so you are customizing the "private" templates, didn't know that was possible.
> > > You can just delete it, it doesn't exist anymore after this change.
> > 
> > 
> > well that's too bad because I'm doing a little customization in there, such as using relative `/img` references.
> 
> I think you could find other way to do that on Gitea.

yeah, the expression will have to be replaced on the proxy.
